### PR TITLE
fix(optimizer)!: replace find / find_all with scoped lookups

### PR DESCRIPTION
**Pull request by fivetran-kwoodbeck**
There are various instances where `find` / `find_all` are used (inappropriately) and search the whole subtree and crossing boundaries, which gives matches to nodes inside nested subqueries.  This leads to wrong decisions, invalid SQL, crashes, etc.

This PR swaps a collection of those calls for scope bounded `find_in_scope` / `find_all_in_scope`, which stop at subquery boundaries.  Each of the changes has a confirmed test that fails on main.

### CODE REVIEW

File: sqlglot/generators/snowflake.py [hunk 1]
```diff
@@ -684,7 +684,7 @@ def with_properties(self, properties: exp.Properties) -> str:
         return self.properties(properties, wrapped=False, prefix=self.sep(""), sep=" ")
 
     def values_sql(self, expression: exp.Values, values_as_table: bool = True) -> str:
-        if expression.find(*self.UNSUPPORTED_VALUES_EXPRESSIONS):
+        if find_in_scope(expression, *self.UNSUPPORTED_VALUES_EXPRESSIONS):
```
Comment (georgesittas) [inline sqlglot/generators/snowflake.py]: This introduces a regression, try the following in main vs in this PR and run it against duckdb, snowflake:

```python
from sqlglot import transpile
sql = "SELECT * FROM (VALUES (1, (SELECT {'k': 1}))) AS v(a, b)"
print(transpile(sql, read="duckdb", write="snowflake")[0])
```

File: sqlglot/optimizer/unnest_subqueries.py [hunk 2]
```diff
@@ -212,7 +212,7 @@ def decorrelate(select, parent_select, external_columns, next_alias_name):
     # if the value of the subquery is not an agg or a key, we need to collect it into an array
     # so that it can be grouped. For subquery projections, we use a MAX aggregation instead.
     agg_func = exp.Max if is_subquery_projection else exp.ArrayAgg
-    if not value.find(exp.AggFunc) and value.this not in group_by:
+    if not find_in_scope(value, exp.AggFunc) and value.this not in group_by:
```
Comment (georgesittas) [inline sqlglot/optimizer/unnest_subqueries.py]: This also looks problematic. Compare main vs the PR on:

```python
import duckdb

from sqlglot import parse_one
from sqlglot.optimizer.normalize import normalize
from sqlglot.optimizer.pushdown_projections import pushdown_projections
from sqlglot.optimizer.qualify import qualify
from sqlglot.optimizer.unnest_subqueries import unnest_subqueries

sql = """
SELECT x.a
FROM x
WHERE x.a = (
    SELECT (SELECT SUM(z.c) FROM z)
    FROM y
    WHERE y.b = x.a
)
"""

schema = {
    "x": {"a": "INT"},
    "y": {"b": "INT"},
    "z": {"c": "INT"},
}

expression = qualify(
    parse_one(sql, read="duckdb"),
    dialect="duckdb",
    schema=schema,
)
expression = pushdown_projections(
    expression,
    schema=schema,
    dialect="duckdb",
)
expression = normalize(expression)
expression = unnest_subqueries(expression)

transformed_sql = expression.sql("duckdb")
print(transformed_sql)

connection = duckdb.connect()
connection.execute(
    """
    CREATE TABLE x(a INT);
    CREATE TABLE y(b INT);
    CREATE TABLE z(c INT);

    INSERT INTO x VALUES (1), (2);
    INSERT INTO y VALUES (1), (2);
    INSERT INTO z VALUES (1), (2);
    """
)

print("Original:", connection.execute(sql).fetchall())
print("Transformed:", connection.execute(transformed_sql).fetchall())
```

**Comment by georgesittas** [comment]
Let's make sure we systematically test for edge cases in these places, especially in the optimizer. If something doesn't look trivially / provably safe, then we can just leave it as-is.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 2 above)
Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/unnest_subqueries.py]: Good catch, but this was patched because `find` was (also) found to produce invalid SQL.  Given they both fail, maybe this is a separate PR?

```
import duckdb
from sqlglot import parse_one
from sqlglot.optimizer import optimize

SQL = (
    "SELECT a FROM t1 "
    "WHERE a IN (SELECT x + (SELECT MAX(z) FROM t3 WHERE z <> t2.x) "
    "            FROM t2 WHERE t2.k = t1.k)"
)
SCHEMA = {"t1": {"a": "int", "k": "int"},
          "t2": {"x": "int", "k": "int"},
          "t3": {"z": "int"}}

optimized = optimize(parse_one(SQL), schema=SCHEMA).sql("duckdb")

con = duckdb.connect()
con.execute("CREATE TABLE t1(a INT, k INT); INSERT INTO t1 VALUES (11,1),(21,1),(5,2)")
con.execute("CREATE TABLE t2(x INT, k INT); INSERT INTO t2 VALUES (1,1),(2,1),(5,2)")
con.execute("CREATE TABLE t3(z INT);        INSERT INTO t3 VALUES (10),(20)")

original_result = con.execute(SQL).fetchall()

print("Optimized SQL:\n ", optimized, "\n")
print("Original result:", original_result)
try:
    optimized_result = con.execute(optimized).fetchall()
    print("Optimized result:", optimized_result)
    assert optimized_result == original_result, "WRONG RESULTS: optimizer changed the answer"
    print("\nPASS: optimizer preserved semantics")
except duckdb.Error as e:
    print("Optimized result: DuckDB ERROR:", str(e).splitlines()[0])
```

**Comment by fivetran-kwoodbeck** [comment]
> Let's make sure we systematically test for edge cases in these places, especially in the optimizer. If something doesn't look trivially / provably safe, then we can just leave it as-is.

These are all cases that have been "proven" to be wrong (at least according to what's in the test submodule).  Each swap has a test case that fails on main and is fixed here.  The first one you found (snowflake) did have an incorrect assertion.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 2 above)
Comment (georgesittas) [reply sqlglot/optimizer/unnest_subqueries.py]: I see, will think about this one a bit more tomorrow, ty.

Comment (georgesittas) [reply sqlglot/optimizer/unnest_subqueries.py]: I think we want something like this:

```python
if (
    not isinstance(value, exp.Subquery)
    and not find_in_scope(value, exp.AggFunc)
    and value.this not in group_by
):
    ...
```

A scalar subquery already returns a single value per group and wrapping `value.this` in an aggregation produces invalid SQL (`ARRAY_AGG(SELECT ...)`).

Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/unnest_subqueries.py]: that works, thanks!

**Review by georgesittas — APPROVED** [review]
